Stream io

thed636 opened the pull request:
Binary serialization via std::ostream interface.

Commit:
Change reading streambuf implementation

Simplify implementation using internal buffer pointers of std::streambuf.

Commit:
Little optimization for single byte values read

Commit:
Remove redundant includes

Commit:
Add support for std::vector<char> to deserialization

std::vector<char> is treated as byta pg type which is raw bytes
array. So the special deserialization is needed.

Commit:
Refactor binary serialization system

Now it uses ostream interface for data streming and
same architecture as binary deserialization.

elsid commented on include/ozo/impl/write.h:
```diff
@@ -0,0 +1,72 @@
+
+template <typename T>
+inline Require<Integral<T> && sizeof(T) == 1, ostream&>
+write(ostream& out, T in) {
```
Use auto for return type, don't put function name on a new line:
```c++
inline auto write(ostream& out, T in)
    -> Require<Integral<T> && sizeof(T) == 1, ostream&> {
```

thed636 replied:
Well, to be honest, I don't remember any conventions about this. At least we have same style in "connection.h", "binary_deserealization.h" and so on. So if you have a suggestion to use this style of return type with Require statement emulation the best choice is to tell us why it is better and get a convention between all of us.

elsid replied:
I don't have any good reason for my suggestion now.

thed636 replied:
For now I discovered the only one good reason - in case of something like this:

```c++
template <typename T1, typename T2>
inline auto foo(very_long<T1, and<sopthisticated_template_exprassion<T>>>& v) -> decltype(bar(v)) {
    return bar(v);
}
```

elsid commented on include/ozo/impl/write.h:
```diff
@@ -0,0 +1,72 @@
+}
+
+template <typename T>
+inline Require<FusionSequence<T>&&!HanaSequence<T>&&!HanaStruct<T>, ostream&>
```
Put spaces between operands and operators to make more readable.

thed636 replied:
**DONE**

elsid commented on include/ozo/detail/ostream.h:
```diff
@@ -0,0 +1,31 @@
+
+namespace ozo::detail {
+
+using std::ostream;
```
Why?

thed636 replied:
Why std::ostream or what? If you are asking about why alias (or using) is needed - the answer is - in the future, I think we will need our own stream or extensions ~~e.g., for strings on which we can make move semantic optimisation, since no byte order conversion is need.~~ (well, strings is not good example - this customisation point is in the different place).

elsid replied:
But why we can't add it in the future? Interface should be the same with std::ostream.

thed636 replied:
Sure, we can. I just did the same thing as for istream.h. The idea was to reserve place for custom stream implementation if it will be needed.

elsid replied:
If you want to do this to restrict clients from use of std::ostream in their code, then better idea is to make adaptor type. Because now code with std::ostream is still correct for compiler, therefore when you will change this type later, clients could suffer.

thed636 replied:
**Done**

elsid commented on include/ozo/binary_query.h:
```diff
@@ -67,13 +67,15 @@ class binary_query {
         std::string_view text;
         buffer_type buffer;
+        ozo::detail::ostreambuf ostreambuf;
+        ozo::ostream stream;
```
Isn't it suppose to be defined somewhere else? This type should contain only what is required to make request using libpq, not to construct it. Streambuf and stream interface is not required to make request.

I can suggest divide field_proxy into field_meta_proxy and field_value_proxy and put stream into field_meta_proxy.
```c++
    template <std::size_t field>
    class field_meta_proxy {
    public:
        field_meta_proxy(impl_type& result, std::ostream& stream)
            : result(result), stream_(stream) {}

        constexpr void set_type(::Oid value) noexcept;
        constexpr void set_format(int value) noexcept;
        constexpr void set_length(int value) noexcept;
        constexpr auto& buffer() noexcept;
        constexpr auto& stream() noexcept {
            return stream_;
        }

    private:
        impl_type& result;
        std::ostream& stream_;
    };
    template <std::size_t field>
    class field_value_proxy {
    public:
        field_value_proxy(impl_type& result) : result(result) {}

        constexpr void set_value_by_offset(const std::size_t offset) noexcept {
            result.values[field] = std::data(result.buffer) + offset;
        }
        constexpr void set_null_value() noexcept {
            result.values[field] = nullptr;
        }

    private:
        impl_type& result;
    };
```
then make_impl should look like this:
```c++
static constexpr auto make_impl(...) {
        // ...
        std::streambuf buffer(result->buffer);
        std::ostream stream(&buffer);
        // ...
                field_meta_proxy<field> proxy(*result, stream);
        // ...
                    field_value_proxy<field>(*result).set_value_by_offset(offset);
                    // ...
                    field_value_proxy<field>(*result).set_null_value();
        // ...
}
```

thed636 replied:
Thanks for the suggestion, I'll try to implement it my best!

Commit:
Fix dangling reference in binary query

Binary query holds text as std::string_view object. This is
correct for static texts like boost::hana strings and so on,
but not suitable for queries made from std::string. In this
case correctness of the string_view depends on string lifetime.

The best solution is to hold text as it-s generic type. The
neccesary modifications have been made.

Perameters passing as hana::tuple significantly simplify code
without any inconvenience for the user API.

Function to_const_char is provided to convert supported types
of query text into native supported const char*.

elsid commented on include/ozo/binary_serialization.h:
```diff
@@ -2,107 +2,58 @@
+};
+
+template <class M, class In>
+inline std::ostream& send(std::ostream& out, const oid_map_t<M>& oid_map, const In& in) {
```
Replace std::ostream -> ostream.

thed636 replied:
**Done**
